Resolve reimbursement tax calculation discrepancies

benjaminwil opened the pull request:
What is the goal of this PR?
---

While we were manually testing the extension for the 1.0 release, we discovered that going through the stock Solidus RMA -> customer return -> reimbursement flow would result in reimbursements being generated with the wrong amount of tax.

After the manual testing session, I discovered that the refund feature test was being skipped. I refactored it so we can more easily test two scenarios:

  1. Removing an entire line item results in the correct reimbursement amount being generated.
  2. Removing part of a line item (1 of 2 quantity) results in the correct reimbursement being generated.

Merge Checklist
---

- [x] Update the changelog

benjaminwil committed:
Create shared context for checkout-related specs

Our checkout and post-checkout feature specs share some `before` setup,
and I wanted to extract it into a shared context to make the
spec-specific set up more clear.

benjaminwil committed:
Refactor skipped refund feature spec

While trying to make this spec pass and investigate an issue with
tax calculations we found during manual testing, I wanted to refactor
this spec so that:

1. It would be more clear which line items we were reimbursing for.
2. It would cover the case where a line item would be removed as well as
   the case where a line item would not be removed but its quantity
   would be changed

This test is pending as the reimbursement amounts populated are
currently wrong ($10.00 instead of $10.89):

    Failures:

      1) Refunding an order adds tax calculated by TaxJar to the order total
         Failure/Error: expect(find(".reimbursement-refund-amount")).to have_content("$10.89")
           expected to find text "$10.89" in "$10.00"
         # ./spec/features/spree/admin/refund_spec.rb:72:in `block (2 levels) in <top (required)>'

I removed the VCR cassette because we will need to re-record it as part
of making the test pass.

benjaminwil committed:
Fix tax calculation for orders with returned items

Before this commit, we were excluding returned and cancelled inventory
units in two circumstances:

  1. For tax calculations
  2. For TaxJar's transaction reporting dashboard

But this was incorrect. To explain this, we must look at orders with
customer returns on them.

We want to exclude returned and cancelled inventory units for reporting
purposes (2), but we must continue calculating tax for any returned or
cancelled units (1). The reason we must continue calculating tax is so
that we have a record of how much tax must be reimbursed to customers
when we generate new `Spree::Reimbursement`s from the admin.

A bit more context:

Before this change, whenever `order.recalculate` is called, we would
recalculate taxes, excluding returned or cancelled items. This in a
returned line item's `#additional_tax_total` being fully zeroed out. A
partially returned line item's tax total would be in an even
harder-to-understand state.

Co-authored-by: Adam Mueller <adam@super.gd>
Co-authored-by: Noah Silvera <noah@super.gd>

benjaminwil commented:
@forkata @Noah-Silvera Just want to update you on the status of this draft. (I will be away for the rest of the week.)

The commit "Fix tax calculation for orders with returned items" fixes the issue where bad reimbursements are generated during the RMA -> reimbursement admin flow. Check out the commit message for more info about 

But there is still something not right. If you examine the `refund.yml` cassette there are 400 responses coming back from the reporting API endpoints because the totals don't line up.

I haven't had time to delve in yet. I think the change Adam and I made is very correct but maybe something is still missing to account for the reporting API requests.

forkata commented on lib/super_good/solidus_taxjar/api_params.rb:
```diff
@@ -111,46 +113,58 @@ def order_address_params(address)
         def line_items_params(line_items)
           {
-            line_items: valid_line_items(line_items).map do |line_item|
+            line_items: line_items.filter_map { |line_item|
```
💡 TIL `filter_map`!

forkata commented on spec/super_good/solidus_taxjar/api_params_spec.rb:
```diff
@@ -221,27 +221,6 @@
       end
     end
 
-    context "when some line items have been returned" do
```
WDYT, about leave this test as a regression spec to ensure that returned/cancelled items are not excluded?

forkata reviewed (commented):
Great work @benjaminwil @adammathys!

forkata commented on spec/fixtures/cassettes/features/spree/admin/refund.yml:
```diff
@@ -0,0 +1,572 @@
+    uri: https://api.taxjar.com/v2/transactions/orders
+    body:
+      encoding: ASCII-8BIT
+      string: '{"customer_id":"1","to_country":"US","to_zip":"11430","to_city":"Herndon","to_state":"NY","to_street":"A
+        Different Road","line_items":[{"id":2,"quantity":1,"product_identifier":"SKU-2","description":"Second
+        Product - Master","product_tax_code":"TaxCode - 597842","unit_price":"10.0","discount":0,"sales_tax":"0.89"},{"id":3,"quantity":2,"product_identifier":"SKU-3","description":"Third
+        Product - Master","product_tax_code":"TaxCode - 597842","unit_price":"10.0","discount":0,"sales_tax":"1.78"}],"transaction_id":"R525233498","transaction_date":"2024-01-23T20:27:39Z","amount":"40.0","shipping":"0.0","sales_tax":"3.56"}'
```
It looks like we're not accounting for the reduced quantity in the `amount` field on the transaction here. This could be related to the fact that the event we hook into is fired before `persist_totals` is called, and thus giving us a false positive failure in the specs.

Noah-Silvera commented:
Just rebased to get the PR that adds a delay in here!

Noah-Silvera committed:
Pend spec failing on CI

This spec is intermittently failing on CI, and it's incredibly hard to
reproduce why locally.

We need to keep moving so pending this on CI for now

Co-authored-by: benjamin wil <benjamin@super.gd>

benjaminwil commented on spec/features/spree/admin/refund_spec.rb:
```diff
@@ -105,6 +105,9 @@
     # amount.
     #
+    if ENV["CI"]
+      pending "A very annoying and hard to debug issue is causing this check to fail consistently in CI, but not locally"
```
If you care to tweak this... it is failing _regularly_ but not _consistently_. If you look at the CI runs it sometimes passes up to 2 of 4 Solidus versions.

Noah-Silvera replied:
God that's so annoying!

Noah-Silvera committed:
Update CHANGELOG.md
